src/Microsoft.PowerShell.Commands.Management/commands/management/Clipboard.cs

```diff
@@ -157,46 +157,46 @@ public static void SetRtf(string plainText, string rtfText)
+        [LibraryImport("kernel32.dll")]
+        private static partial IntPtr GlobalAlloc(uint flags, UIntPtr dwBytes);
 
-        [DllImport("kernel32.dll")]
-        private static extern IntPtr GlobalFree(IntPtr hMem);
+        [LibraryImport("kernel32.dll")]
+        private static partial IntPtr GlobalFree(IntPtr hMem);
 
-        [DllImport("kernel32.dll")]
-        private static extern IntPtr GlobalLock(IntPtr hMem);
+        [LibraryImport("kernel32.dll")]
+        private static partial IntPtr GlobalLock(IntPtr hMem);
 
-        [DllImport("kernel32.dll")]
+        [LibraryImport("kernel32.dll")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        private static extern bool GlobalUnlock(IntPtr hMem);
+        private static partial bool GlobalUnlock(IntPtr hMem);
 
-        [DllImport("kernel32.dll", ExactSpelling = true, EntryPoint = "RtlMoveMemory", SetLastError = true)]
-        private static extern void CopyMemory(IntPtr dest, IntPtr src, uint count);
+        [LibraryImport("kernel32.dll", EntryPoint = "RtlMoveMemory")]
+        private static partial void CopyMemory(IntPtr dest, IntPtr src, uint count);
 
-        [DllImport("user32.dll", SetLastError = false)]
+        [LibraryImport("user32.dll")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        private static extern bool IsClipboardFormatAvailable(uint format);
+        private static partial bool IsClipboardFormatAvailable(uint format);
 
-        [DllImport("user32.dll", SetLastError = true)]
+        [LibraryImport("user32.dll")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        private static extern bool OpenClipboard(IntPtr hWndNewOwner);
+        private static partial bool OpenClipboard(IntPtr hWndNewOwner);
 
-        [DllImport("user32.dll", SetLastError = true)]
+        [LibraryImport("user32.dll")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        private static extern bool CloseClipboard();
+        private static partial bool CloseClipboard();
 
-        [DllImport("user32.dll", SetLastError = true)]
+        [LibraryImport("user32.dll")]
         [return: MarshalAs(UnmanagedType.Bool)]
-        private static extern bool EmptyClipboard();
+        private static partial bool EmptyClipboard();
 
-        [DllImport("user32.dll", SetLastError = true)]
-        private static extern IntPtr GetClipboardData(uint format);
+        [LibraryImport("user32.dll")]
+        private static partial IntPtr GetClipboardData(uint format);
 
-        [DllImport("user32.dll")]
-        private static extern IntPtr SetClipboardData(uint format, IntPtr data);
+        [LibraryImport("user32.dll")]
+        private static partial IntPtr SetClipboardData(uint format, IntPtr data);
 
-        [DllImport("user32.dll", SetLastError = true)]
-        private static extern uint RegisterClipboardFormat(string lpszFormat);
+        [LibraryImport("user32.dll", StringMarshalling = StringMarshalling.Utf16)]
```

Comment:
There was no `CharSet = CharSet.Unicode` setting before, so why `StringMarshalling.Utf16` is needed in this case?

Reply:
Marshalling of string is apparently not supported without explicit marshalling information. 

https://github.com/dotnet/runtime/blob/9e8c7a8f65ad86ee68d316de4e3a5deea72b7edd/docs/design/libraries/LibraryImportGenerator/Compatibility.md#string-marshalling

Reply:
When `DllImportAttribute.CharSet` is not set, it is up to the language compiler to determine which CharSet to use. C#, uses the [None](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.charset#system-runtime-interopservices-charset-none) charset by default, which has the same behavior as the [Ansi](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.charset#system-runtime-interopservices-charset-ansi) charset.

https://learn.microsoft.com/en-us/dotnet/standard/native-interop/charset

Reply:
Thanks @xtqqczze. It's called out more clearly [here](https://github.com/dotnet/runtime/blob/9e8c7a8f65ad86ee68d316de4e3a5deea72b7edd/docs/design/libraries/LibraryImportGenerator/Compatibility.md#semantic-changes-compared-to-dllimportattribute):
> The P/Invoke source generator makes no assumptions about StringMarshalling if it is not explicitly set on LibraryImportAttribute. Marshalling of char or string requires explicitly specifying marshalling information.

Then another question is, how to select between utf8 and utf16? Should we always use utf16 on Windows and utf8 on Unix platforms?

Reply:
This depends on target API. On Windows most of APIs use UTF16, but LDAP uses UTF8 as per the standard, also BER string is used sometimes (crypto API?).